include/lsst/afw/typehandling/HeteroMap.h

```diff
@@ -0,0 +1,700 @@
+// -*- LSST-C++ -*-
+/*
+ * This file is part of afw.
+ *
+ * Developed for the LSST Data Management System.
+ * This product includes software developed by the LSST Project
+ * (https://www.lsst.org).
+ * See the COPYRIGHT file at the top-level directory of this distribution
+ * for details of code ownership.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+#ifndef LSST_AFW_TYPEHANDLING_HETEROMAP_H
+#define LSST_AFW_TYPEHANDLING_HETEROMAP_H
+
+#include <functional>
+#include <ostream>
+#include <memory>
+#include <sstream>
+#include <typeinfo>
+#include <type_traits>
+#include <unordered_set>
+#include <vector>
+
+#include "boost/core/demangle.hpp"
+#include "boost/variant.hpp"
+
+#include "lsst/pex/exceptions.h"
+#include "lsst/afw/typehandling/Storable.h"
+
+namespace lsst {
+namespace afw {
+namespace typehandling {
+
+/**
+ * Key for type-safe lookup in a HeteroMap.
+ *
+ * @tparam K the logical type of the key (e.g., a string)
+ * @tparam V the type of the value mapped to this key
+ *
+ * Key objects are equality-comparable, hashable, sortable, or printable if and only if `K` is comparable,
+ * hashable, sortable, or printable, respectively. Key can be used in compile-time expressions if and only
+ * if `K` can (in particular, `Key<std::string, V>` cannot).
+ *
+ * @note Objects of this type are immutable.
+ */
+template <typename K, typename V>
+class Key final {
+public:
+    using KeyType = K;
+    using ValueType = V;
+
+    /**
+     * Construct a new key.
+     *
+     * @param name
+     *            the name of the field. For most purposes, this value is
+     *            the actual key; it can be retrieved by calling getId().
+     *
+     * @exceptsafe Provides the same exception safety as the copy-constructor of `K`.
+     *
+     * @see makeKey
+     */
+    constexpr Key(K name) : id(name) {}
+
+    Key(Key const&) = default;
+    Key(Key&&) = default;
```

Comment:
Yes, `default` makes it `noexcept` if the member constructors are. Also, after I made a mess of `geom` I want to avoid making methods conditionally `noexcept`; the syntax is just too user-unfriendly.